### ISSUE TITLE
Upload images in batches will cause images to repeat

### ISSUE THREAD

**Issue by sunosv** [LycheeOrg/Lychee#903]
### Detailed description of the problem [REQUIRED]
When images are uploaded in batches, using a browser There's a chance that two images will repeat. 
The two images are not the same images before uploading.

### Steps to reproduce the issue
Upload images in batches

**Screenshots**
![https://i.loli.net/2021/02/02/UxCXnPKslWqIyAV.png](https://i.loli.net/2021/02/02/UxCXnPKslWqIyAV.png)

### Output of the diagnostics  [REQUIRED]
*(Settings => Diagnostics or https://example.com/Diagnostics or `php artisan lychee:diagnostics`)*
`Diagnostics
    -------
    Warning: Dropbox import not working. dropbox_key is empty.

    System Information
    --------------
    Lychee Version (release):        4.2.1
    DB Version:                      4.2.1
    
    composer install:                --no-dev
    APP_ENV:                         production
    APP_DEBUG:                       true
    
    System:                          WINNT
    PHP Version:                     7.4
    Max uploaded file size:          10000M
    Max post size:                   10000M
    MySQL Version:                   5.7.26
    
    Imagick:                         1
    Imagick Active:                  1
    Imagick Version:                 1799
    GD Version:                      bundled (2.1.0 compatible)

    Config Information
    --------------
    version:                         040201
    check_for_updates:               0
    sorting_Photos_col:              title
    sorting_Photos_order:            ASC
    sorting_Albums_col:              title
    sorting_Albums_order:            ASC
    imagick:                         1
    skip_duplicates:                 1
    small_max_width:                 0
    small_max_height:                360
    medium_max_width:                1920
    medium_max_height:               1080
    lang:                            en
    layout:                          1
    image_overlay:                   1
    image_overlay_type:              desc
    default_license:                 none
    compression_quality:             90
    full_photo:                      1
    delete_imported:                 0
    Mod_Frame:                       1
    Mod_Frame_refresh:               30
    thumb_2x:                        1
    small_2x:                        1
    medium_2x:                       1
    landing_page_enable:             0
    landing_owner:                   John Smith
    landing_title:                   John Smith
    landing_subtitle:                Cats, Dogs & Humans Photography
    landing_facebook:                https://www.facebook.com/JohnSmith
    landing_flickr:                  https://www.flickr.com/JohnSmith
    landing_twitter:                 https://www.twitter.com/JohnSmith
    landing_instagram:               https://instagram.com/JohnSmith
    landing_youtube:                 https://www.youtube.com/JohnSmith
    landing_background:              dist/cat.jpg
    site_title:                      Lychee v4
    site_copyright_enable:           1
    site_copyright_begin:            2019
    site_copyright_end:              2019
    additional_footer_text:          
    display_social_in_gallery:       0
    public_search:                   0
    SL_enable:                       0
    SL_for_admin:                    0
    public_recent:                   0
    recent_age:                      1
    public_starred:                  0
    downloadable:                    0
    photos_wraparound:               1
    map_display:                     0
    zip64:                           1
    map_display_public:              0
    map_provider:                    Wikimedia
    force_32bit_ids:                 0
    map_include_subalbums:           0
    update_check_every_days:         3
    has_exiftool:                    1
    share_button_visible:            0
    import_via_symlink:              0
    has_ffmpeg:                      1
    location_decoding:               0
    location_decoding_timeout:       30
    location_show:                   1
    location_show_public:            0
    rss_enable:                      0
    rss_recent_days:                 7
    rss_max_items:                   100
    prefer_available_xmp_metadata:   0
    editor_enabled:                  1
    lossless_optimization:           0
    swipe_tolerance_x:               150
    swipe_tolerance_y:               250
    local_takestamp_video_formats:   .avi|.mov
    log_max_num_line:                1000
    unlock_password_photos_with_url_param: 0
    nsfw_visible:                    1
    nsfw_blur:                       0
    nsfw_warning:                    0
    nsfw_warning_admin:              0`
### Browser and system
Chrome

**Comment by ildyria** [comment]
My guess is that the name generation which is based on the time of upload is colliding.

**Comment by sunosv** [comment]
> My guess is that the name generation which is based on the time of upload is colliding.

I would be very grateful if this problem could be fixed.

**Comment by kamil4** [comment]
This looks Windows-specific. Probably `microtime` result is not updated between every call. It could be this bug: https://bugs.php.net/bug.php?id=64633

We could try combining the result of `microtime` with `hrtime` and see if that helps.

A more robust fix would be to verify that the generated filename prefix is not already in use. I.e., check early on for the existence of `thumb/<prefix>.jpeg`; if found, sleep for a random fraction of a second and try regenerating the prefix again. Given the inherently race'y nature of the problem at hand, the test would have to be atomic (something like `open(..., O_CREAT | O_EXCL)`).

**Comment by kamil4** [comment]
Another way would be to limit the number of uploads allowed to be processed at a time. Currently there's no limit, and I think that can already cause unrelated problems on systems with fast upload speed, where the server could run out of memory processing too many files at a time. I've definitely seen it with GD as the image processor...

I think it would be easy for the front end to ensure that the upload of the next file does not begin until the number of uploads being processed falls below a threshold (which could be a config variable). With a threshold of `1` things would _probably_ be safe, and with `0` (i.e., _don't start uploading the next one until there are no files being processed_) they would definitely be.

**Comment by sunosv** [comment]
> Another way would be to limit the number of uploads allowed to be processed at a time. Currently there's no limit, and I think that can already cause unrelated problems on systems with fast upload speed, where the server could run out of memory processing too many files at a time. I've definitely seen it with GD as the image processor...
> 
> I think it would be easy for the front end to ensure that the upload of the next file does not begin until the number of uploads being processed falls below a threshold (which could be a config variable). With a threshold of `1` things would _probably_ be safe, and with `0` (i.e., _don't start uploading the next one until there are no files being processed_) they would definitely be.
Will the limits affect upload speed？ And with hrtime ,how to do it.

**Comment by kamil4** [comment]
> Will the limits affect upload speed?

Somewhat. It would, in some cases, introduce gaps between uploads (to allow the server to "catch up"). Upload speeds of individual files would not be affected.

> And with hrtime ,how to do it.

OK, here's a quick proof-of-concept you can try. You are on the 4.2.1 release, correct? You need to locate the following line:

https://github.com/LycheeOrg/Lychee/blob/a368964afe5b7b9403ca97a1b7959f923e2f0b28/app/Actions/Photo/Create.php#L83

And replace it with this line:
```
        $this->photo_Url = md5(microtime() . strval(hrtime(true))) . $this->extension;
```
Let us know if that fixes the issue you are seeing.

**Comment by sunosv** [comment]
> > Will the limits affect upload speed?
> 
> Somewhat. It would, in some cases, introduce gaps between uploads (to allow the server to "catch up"). Upload speeds of individual files would not be affected.
> 
> > And with hrtime ,how to do it.
> 
> OK, here's a quick proof-of-concept you can try. You are on the 4.2.1 release, correct? You need to locate the following line:
> 
> https://github.com/LycheeOrg/Lychee/blob/a368964afe5b7b9403ca97a1b7959f923e2f0b28/app/Actions/Photo/Create.php#L83
> 
> And replace it with this line:
> 
> ```
>         $this->photo_Url = md5(microtime() . strval(hrtime(true))) . $this->extension;
> ```
> 
> Let us know if that fixes the issue you are seeing.

Bed news，It still the same. I tested it 4 times. The issue appeared for the third time.

**Comment by kamil4** [comment]
> Bed news，It still the same. I tested it 4 times. The issue appeared for the third time.

OK, so the quick hack didn't work. We'll need to do it properly then. This is my Plan B:

> A more robust fix would be to verify that the generated filename prefix is not already in use. I.e., check early on for the existence of thumb/<prefix>.jpeg; if found, sleep for a random fraction of a second and try regenerating the prefix again. Given the inherently race'y nature of the problem at hand, the test would have to be atomic (something like open(..., O_CREAT | O_EXCL)).

It may take a couple of days before I get to it though. I have some PRs waiting for a review...

**Comment by sunosv** [comment]
> It may take a couple of days before I get to it though. I have some PRs waiting for a review...

Ok , I'm looking forward.

**Comment by alex-phillips** [comment]
Not sure if this is related, if not I can open a new issue. But it seems every time I use the `lychee:sync` command, images are being duplicated here as well. It's as though lychee is no longer detecting that these images already exist and matching them up with what has previously been synced.

**Comment by kamil4** [comment]
> Not sure if this is related, if not I can open a new issue. But it seems every time I use the `lychee:sync` command, images are being duplicated here as well.

No, it's not related. This issue is about race conditions during parallel uploads. That can't be an issue with imports (which are always done sequentially).

You should open a new issue. I am surprised by your report though as I've been hacking on the import code in #894 and I know that skipping duplicates _does_ work, at least in simple cases.

**Comment by kamil4** [comment]
I take it back, @alex-phillips. I was able to reproduce what you report on `master`. Still, this code has been refactored in #894 and once that's merged the problem should be fixed.

**Comment by sunosv** [comment]
> It may take a couple of days before I get to it though. I have some PRs waiting for a review...

Hey~ How is it going with this issue?

**Comment by kamil4** [comment]
Working on it as we speak...

**Comment by sunosv** [comment]
> Working on it as we speak...

Thank you!

**Comment by kamil4** [comment]
OK, I've got the fix: #907.

If you want to test it without using git, note that the patch may not apply cleanly to the 4.2.1 release due to extensive changes to `app/Actions/Photo/Rotate.php` made since the release.

But, for testing purposes, you can skip that file. Simply create a new file `app/Actions/Photo/Extensions/Prefix.php` with the content from https://raw.githubusercontent.com/LycheeOrg/Lychee/9da51ec4cb48337f11f2ebc16e06c015f4bf5ac5/app/Actions/Photo/Extensions/Prefix.php, and edit the three lines in `app/Actions/Photo/Create.php` as shown in https://github.com/LycheeOrg/Lychee/pull/907/files#diff-685f47499dc041e9e6adcdfc1f9dfc502beb845b0c509570dfeeb628acb0e123

**Comment by kamil4** [comment]
I just added one more tweak, to ` app/Actions/Photo/Strategies/StrategyDuplicate.php`. You should be able to apply it to the 4.2.1 release, but for simple testing you can ignore it.

**Comment by sunosv** [comment]
> OK, I've got the fix: #907.
> 
> If you want to test it without using git, note that the patch may not apply cleanly to the 4.2.1 release due to extensive changes to `app/Actions/Photo/Rotate.php` made since the release.
> 
> But, for testing purposes, you can skip that file. Simply create a new file `app/Actions/Photo/Extensions/Prefix.php` with the content from https://raw.githubusercontent.com/LycheeOrg/Lychee/9da51ec4cb48337f11f2ebc16e06c015f4bf5ac5/app/Actions/Photo/Extensions/Prefix.php, and edit the three lines in `app/Actions/Photo/Create.php` as shown in https://github.com/LycheeOrg/Lychee/pull/907/files#diff-685f47499dc041e9e6adcdfc1f9dfc502beb845b0c509570dfeeb628acb0e123

Problem solved. Good job!  Thank you~
